src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java

```diff
@@ -0,0 +1,190 @@
+
+import hudson.EnvVars;
+import org.jenkinsci.plugins.workflow.actions.ArgumentsAction;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import static org.jenkinsci.plugins.workflow.actions.ArgumentsAction.NotStoredReason.MASKED_VALUE;
+
+/**
+ * Implements {@link ArgumentsAction} by storing step arguments, with sanitization
+ */
+public class ArgumentsActionImpl extends ArgumentsAction {
+
+    /** Arguments to the step, for cases where we cannot simply store the step because masking was applied */
+    @CheckForNull
+    private Map<String,Object> arguments;
+
+
+    private boolean isModifiedBySanitization = false;
+
+    public boolean isModifiedBySanitization() {
+        return isModifiedBySanitization;
+    }
+
+    /** FIXME needs to handle getting an uninstantiatedDeliverable or a fully configured step instance in the params map */
+    public ArgumentsActionImpl(@Nonnull Map<String, Object> stepArguments, @Nullable EnvVars env) {
+        // TODO add handling with $class syntax and direct Step creation
+        Map<String,Object> sanitizedArguments = sanitizedStepArguments(stepArguments, env);
+        for (Object o : sanitizedArguments.values()) {
+            if (o != null && (o.equals(MASKED_VALUE) || o.equals(NotStoredReason.OVERSIZE_VALUE))) {
+                isModifiedBySanitization = true;
+                break;
+            }
+        }
+        arguments = sanitizedArguments;
+    }
+
+    /** Create a step, sanitizing strings for secured content */
+    public ArgumentsActionImpl(@Nonnull Map<String, Object> stepArguments) {
+        this(stepArguments, new EnvVars());
+    }
+
+    /** See if sensitive environment variable content is in a string */
+    public static boolean isStringSafe(@CheckForNull String input, @CheckForNull EnvVars variables, @Nonnull Set<String> safeEnvVariables) {
+        if (input == null || variables == null || variables.size() == 0) {
+            return true;
+        }
+        StringBuilder pattern = new StringBuilder();
+        int count = 0;
+        for (Map.Entry<String,String> ent : variables.entrySet()) {
+            String val = ent.getValue();
+            if (val == null || val.isEmpty() || safeEnvVariables.contains(ent.getKey())) {  // Skip values that are safe
+                continue;
+            }
+            if (count > 0) {
+                pattern.append('|');
+            }
+            pattern.append(Pattern.quote(val));
+            count++;
+        }
+        return (count > 0)
+                ? !Pattern.compile(pattern.toString()).matcher(input).find()
+                : true;
+    }
+
+    private static final HashSet<String> SAFE_ENVIRONMENT_VARIABLES = new HashSet<String>(Arrays.asList(
```

Comment:
What makes these “safe”? What would make other environment variables “unsafe”? This looks wrong to me, but at least it deserves some careful Javadoc explaining how you arrived at this list and how it would be maintained.